google/cloud/storage/internal/openssl_util.h

```diff
@@ -0,0 +1,245 @@
+    BUF_MEM* buf_mem;
+    BIO_get_mem_ptr(static_cast<BIO*>(bio_chain.get()), &buf_mem);
+    // Return a string copy of the buffer's bytes, as the buffer will be freed
+    // upon this method's exit.
+    return std::string(buf_mem->data, buf_mem->length);
+  }
+
+  /**
+   * Transform a string in-place, removing trailing occurrences of a character.
+   *
+   * Warning: this was written with the intent of operating on a string
+   * containing ASCII-encoded (8-bit) characters (e.g. removing trailing '='
+   * characters from a Base64-encoded string) and may not function correctly
+   * for strings containing Unicode characters.
+   */
+  static void RightTrim(std::string& str, char trim_ch) {
+    if (str.length() == 0) return;
+    size_t end_pos = str.find_last_not_of(trim_ch);
+    if (std::string::npos != end_pos) str.resize(end_pos + 1);
+  }
+
+  /**
+   * Sign a string with the private key from a PEM container.
+   */
+  static std::string SignStringWithPem(
+      std::string const& str, std::string const& pem_contents,
+      google::cloud::storage::internal::JwtSigningAlgorithms alg) {
+    // We check for failures several times, so we shorten this into a lambda
+    // to avoid bloating the code with alloc/init checks.
+    auto func_name = __func__;  // Avoid using the lambda name instead.
+    auto handle_openssl_failure = [&func_name](std::string error_msg) -> void {
+      std::ostringstream err_builder;
+      err_builder << "Permanent error in " << func_name
+                  << " (failed to sign string with PEM key): " << std::endl
+                  << error_msg;
+      google::cloud::internal::RaiseRuntimeError(err_builder.str());
+    };
+
+    auto digest_ctx = GetDigestCtx();
+    if (not(digest_ctx)) {
+      handle_openssl_failure("Could not create context for OpenSSL digest.");
+    }
+
+    const EVP_MD* digest_type = nullptr;
+    switch (alg) {
+      case JwtSigningAlgorithms::RS256:
+        digest_type = EVP_sha256();
+        break;
+    }
+    if (digest_type == nullptr) {
+      handle_openssl_failure("Could not find specified digest in OpenSSL.");
+    }
+
+    auto pem_buffer = std::unique_ptr<BIO, decltype(&BIO_free)>(
+        BIO_new_mem_buf(const_cast<char*>(pem_contents.c_str()),
+                        pem_contents.length()),
+        &BIO_free);
+    if (not(pem_buffer)) handle_openssl_failure("Could not create PEM buffer.");
+
+    auto private_key = std::unique_ptr<EVP_PKEY, decltype(&EVP_PKEY_free)>(
+        PEM_read_bio_PrivateKey(
+            static_cast<BIO*>(pem_buffer.get()),
+            nullptr,  // EVP_PKEY **x
+            nullptr,  // pem_password_cb *cb -- a custom callback.
+            // void *u -- this represents the password for the PEM (only
+            // applicable for formats such as PKCS12 (.p12 files) that use
+            // a password, which we don't currently support.
+            nullptr),
+        &EVP_PKEY_free);
+    if (not(private_key)) {
```

Comment:
s/not(private_key)/not private_key/ maybe?

Reply:
Sure, done.